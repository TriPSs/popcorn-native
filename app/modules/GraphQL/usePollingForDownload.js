import React, { useState, useEffect } from 'react'

import constants from 'modules/constants'

/**
 * Starts polling for a certain download
 *
 * @param download
 * @param downloadManager
 * @returns {unknown}
 */
export const usePollingForDownload = (download, downloadManager) => {
  const [data, setPollingData] = useState(null)

  useEffect(() => {
      if (download && [constants.STATUS_QUEUED, constants.STATUS_DOWNLOADING].includes(download.status)) {
        downloadManager.pollDownload(download, setPollingData)
      }

      return () => {
        if (download) {
          downloadManager.stopPollDownload(download)
        }
      }
    }, [download],
  )

  return data
}

export default usePollingForDownload
