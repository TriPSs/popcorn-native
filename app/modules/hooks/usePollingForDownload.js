import React from 'react'

import constants from 'modules/constants'
import { useDownloadManager } from 'modules/DownloadManager'

/**
 * States that are valid to pull
 * @type {(string)[]}
 */
export const validPollStates = [
  constants.STATUS_QUEUED,
  constants.STATUS_CONNECTING,
  constants.STATUS_DOWNLOADING,
]

/**
 * Starts polling for a certain download
 */
export const usePollingForDownload = ({ _id }) => {
  const downloadManager = useDownloadManager()

  const download = downloadManager.getDownload(_id)

  React.useEffect(() => {
    if (download && validPollStates.includes(download.status)) {
      downloadManager.pollDownload(download)
    }

    return () => {
      if (download) {
        downloadManager.stopPollDownload(download)
      }
    }
  }, [download?._id, download?.status])

  return download
}

export default usePollingForDownload
