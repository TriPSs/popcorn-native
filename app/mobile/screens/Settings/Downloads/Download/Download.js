import React from 'react'
import PropTypes from 'prop-types'
import { View } from 'react-native'

import usePollingForDownload from 'modules/GraphQL/usePollingForDownload'
import dimensions from 'modules/dimensions'
import constants from 'modules/constants'

import BaseButton from 'components/BaseButton'
import Container from 'components/Container'
import Icon from 'components/Icon'
import Typography from 'components/Typography'

export const styles = {

  root: {
    paddingTop: dimensions.UNIT / 2,
    paddingBottom: dimensions.UNIT,

    display: 'flex',
    flexDirection: 'row',
  },

  divider: {
    marginBottom: 1,
  },

  iconContainer: {
    display: 'flex',
    justifyContent: 'center',
    margin: dimensions.UNIT,
  },
}

export const Download = ({ download, downloadManager, refreshScreen }) => {
  const data = usePollingForDownload(download, downloadManager)
  const { status, progress, speed, timeRemaining, numPeers, quality } = data || download
  const { movie, episode } = download

  const removeDownload = async() => {
    await downloadManager.removeDownload({
      ...download,
      // TODO:: Make a util and get the show title from there so format will be:
      // Show: S01E01. Episode Title
      title: movie?.title ?? `${episode?.show?.title}: S${episode?.season}E${episode?.number}. ${episode?.title}`
    })

    refreshScreen()
  }

  return (
    <BaseButton
      onPress={() => downloadManager.onPress(download)}
      onLongPress={removeDownload}
      rippleBorderless={false}>
      <Container
        style={[styles.root, styles.divider]}
        elevation={2}>

        <View style={styles.iconContainer}>
          <Icon name={'movie'} />
        </View>

        <View>
          <Typography variant={'subtitle2'}>
            {movie?.title ?? `${episode?.show?.title}: S${episode?.season}E${episode?.number}. ${episode?.title}`}
          </Typography>

          <Typography variant={'caption'}>
            {quality} - {status}
          </Typography>

          {status === constants.STATUS_DOWNLOADING && (
            <React.Fragment>
              <Typography variant={'caption'}>
                {`${progress}% / ${speed}`}
              </Typography>

              <Typography variant={'caption'}>
                {timeRemaining}
              </Typography>

              <Typography variant={'caption'}>
                {`${numPeers} peer${numPeers > 1 ? 's' : ''}`}
              </Typography>
            </React.Fragment>
          )}

        </View>
      </Container>
    </BaseButton>
  )
}

Download.propTypes = {
  data: PropTypes.object,
}

Download.defaultProps = {
  data: null,
}

export default Download
