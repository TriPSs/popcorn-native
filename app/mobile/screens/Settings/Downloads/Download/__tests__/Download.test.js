import React from 'react'
import { shallow } from 'enzyme'

import usePollingForDownload from 'modules/hooks/usePollingForDownload'

import Download from '../Download'

jest.mock('modules/GraphQL/usePollingForDownload')

describe('(Screens) Settings > Downloads > Download', () => {

  const downloadManager = {
    removeDownload: jest.fn(),
    onPress: jest.fn(),
  }

  it('should render queued correctly for episode', () => {
    const result = shallow(
      <Download
        downloadManager={downloadManager}
        refreshScreen={jest.fn()}
        download={{
          status: 'queued',
          progress: 0,
          speed: 0,
          timeRemaining: 0,
          numPeers: 0,
          movie: null,
          episode: {
            title: 'episode',
            season: 1,
            number: 1,
            show: {
              title: 'show',
            },
          },
        }}
      />,
    )

    expect(result).toMatchSnapshot()
  })

  it('should render queued correctly for movie', () => {
    const result = shallow(
      <Download
        downloadManager={downloadManager}
        refreshScreen={jest.fn()}
        download={{
          status: 'queued',
          progress: 0,
          speed: 0,
          timeRemaining: 0,
          numPeers: 0,
          movie: {
            title: 'movie',
          },
        }}
      />,
    )

    expect(result).toMatchSnapshot()
  })

  it('should render downloading correctly for movie', () => {
    const result = shallow(
      <Download
        downloadManager={downloadManager}
        refreshScreen={jest.fn()}
        download={{
          status: 'downloading',
          progress: 10,
          speed: '50 kbs',
          timeRemaining: '5 minutes',
          numPeers: 6,
          movie: {
            title: 'movie',
          },
        }}
      />,
    )

    expect(result).toMatchSnapshot()
  })


})
