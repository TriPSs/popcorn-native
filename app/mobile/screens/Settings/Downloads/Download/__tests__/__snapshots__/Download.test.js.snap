// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Screens) Settings > Downloads > Download should render downloading correctly for movie 1`] = `
<BaseButton
  innerRef={null}
  onLongPress={[Function]}
  onPress={[Function]}
  rippleBorderless={false}
  rippleColor="rgba(0, 0, 0, .3)"
>
  <Container
    elevation={2}
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "row",
          "paddingBottom": 8,
          "paddingTop": 4,
        },
        Object {
          "marginBottom": 1,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "display": "flex",
          "justifyContent": "center",
          "margin": 8,
        }
      }
    >
      <Icon
        color="white"
        emphasis="high"
        name="movie"
        size={24}
        style={null}
      />
    </View>
    <View>
      <Typography
        color="white"
        component={null}
        emphasis="high"
        fontWeight="default"
        style={null}
        textProps={Object {}}
        variant="subtitle2"
      >
        movie
      </Typography>
      <Typography
        color="white"
        component={null}
        emphasis="high"
        fontWeight="default"
        style={null}
        textProps={Object {}}
        variant="caption"
      >
        downloading
      </Typography>
      <Typography
        color="white"
        component={null}
        emphasis="high"
        fontWeight="default"
        style={null}
        textProps={Object {}}
        variant="caption"
      >
        10% / 50 kbs
      </Typography>
      <Typography
        color="white"
        component={null}
        emphasis="high"
        fontWeight="default"
        style={null}
        textProps={Object {}}
        variant="caption"
      >
        5 minutes
      </Typography>
      <Typography
        color="white"
        component={null}
        emphasis="high"
        fontWeight="default"
        style={null}
        textProps={Object {}}
        variant="caption"
      >
        6 peers
      </Typography>
    </View>
  </Container>
</BaseButton>
`;

exports[`(Screens) Settings > Downloads > Download should render queued correctly for episode 1`] = `
<BaseButton
  innerRef={null}
  onLongPress={[Function]}
  onPress={[Function]}
  rippleBorderless={false}
  rippleColor="rgba(0, 0, 0, .3)"
>
  <Container
    elevation={2}
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "row",
          "paddingBottom": 8,
          "paddingTop": 4,
        },
        Object {
          "marginBottom": 1,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "display": "flex",
          "justifyContent": "center",
          "margin": 8,
        }
      }
    >
      <Icon
        color="white"
        emphasis="high"
        name="movie"
        size={24}
        style={null}
      />
    </View>
    <View>
      <Typography
        color="white"
        component={null}
        emphasis="high"
        fontWeight="default"
        style={null}
        textProps={Object {}}
        variant="subtitle2"
      >
        show: S1E1. episode
      </Typography>
      <Typography
        color="white"
        component={null}
        emphasis="high"
        fontWeight="default"
        style={null}
        textProps={Object {}}
        variant="caption"
      >
        queued
      </Typography>
    </View>
  </Container>
</BaseButton>
`;

exports[`(Screens) Settings > Downloads > Download should render queued correctly for movie 1`] = `
<BaseButton
  innerRef={null}
  onLongPress={[Function]}
  onPress={[Function]}
  rippleBorderless={false}
  rippleColor="rgba(0, 0, 0, .3)"
>
  <Container
    elevation={2}
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "row",
          "paddingBottom": 8,
          "paddingTop": 4,
        },
        Object {
          "marginBottom": 1,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "display": "flex",
          "justifyContent": "center",
          "margin": 8,
        }
      }
    >
      <Icon
        color="white"
        emphasis="high"
        name="movie"
        size={24}
        style={null}
      />
    </View>
    <View>
      <Typography
        color="white"
        component={null}
        emphasis="high"
        fontWeight="default"
        style={null}
        textProps={Object {}}
        variant="subtitle2"
      >
        movie
      </Typography>
      <Typography
        color="white"
        component={null}
        emphasis="high"
        fontWeight="default"
        style={null}
        textProps={Object {}}
        variant="caption"
      >
        queued
      </Typography>
    </View>
  </Container>
</BaseButton>
`;
